File: t-tasks/src/main/java/com/teo/ttasks/ui/items/TaskItem.java
package com.teo.ttasks.ui.items;

import android.databinding.DataBindingUtil;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.support.graphics.drawable.VectorDrawableCompat;
import android.util.DisplayMetrics;
import android.util.TypedValue;
import android.view.View;

import com.mikepenz.fastadapter.FastAdapter;
import com.mikepenz.fastadapter.IClickable;
import com.mikepenz.fastadapter.IExpandable;
import com.mikepenz.fastadapter.IItem;
import com.mikepenz.fastadapter.ISubItem;
import com.mikepenz.fastadapter.expandable.items.AbstractExpandableItem;
import com.teo.ttasks.R;
import com.teo.ttasks.data.model.TTask;
import com.teo.ttasks.databinding.ItemTaskBinding;
import com.teo.ttasks.util.DateUtils;

import java.util.Comparator;
import java.util.Date;
import java.util.List;

import lombok.Getter;
import lombok.experimental.Accessors;
import timber.log.Timber;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;
import static com.teo.ttasks.util.DateUtils.sdfDay;
import static com.teo.ttasks.util.DateUtils.sdfMonth;

@Accessors()
public class TaskItem<Parent extends IItem & IExpandable & ISubItem & IClickable> extends AbstractExpandableItem<Parent, TaskItem.ViewHolder, TaskItem<Parent>> implements Comparable<TaskItem> {

    /**
     * Comparator that sorts {@link TaskItem}s by their completion date in descending order.
     */
    public static Comparator<TaskItem> completionDateComparator = (lhs, rhs) -> {

        final boolean sameDay = sdfDay.format(lhs.completed).equals(sdfDay.format(rhs.completed));
        final boolean sameMonth = sdfMonth.format(lhs.completed).equals(sdfMonth.format(rhs.completed));

        int returnCode;

        // Check the completed date
        if (lhs.completed == null && rhs.completed == null) {
            returnCode = 0;
        } else if (lhs.completed == null) {
            returnCode = 1;
        } else if (rhs.completed == null) {
            returnCode = -1;
        } else {
            returnCode = rhs.completed.compareTo(lhs.completed);
        }

        // Decide whether the day number and name will be shown i.e. whether the task item will be combined or not
        if (sameDay) {
            if (returnCode > 0) {
                // rhs comes after lhs
                lhs.combineDay = true;
            } else {
                // lhs comes after rhs
                rhs.combineDay = true;
            }
        }

        // Decide whether the month should be shown
        if (sameMonth) {
            if (returnCode > 0) {
                // rhs comes after lhs
                lhs.combineMonth = true;
            } else {
                // lhs comes after rhs
                rhs.combineMonth = true;
            }
        }

        return returnCode;
    };

    public static Comparator<TaskItem> alphabeticalComparator = (lhs, rhs) -> {
        // Titles should never be null or empty but just in case
        final boolean noTitleLeft = lhs.title == null || lhs.title.isEmpty();
        final boolean noTitleRight = rhs.title == null || rhs.title.isEmpty();
        if (noTitleLeft && noTitleRight)
            return 0;
        if (noTitleLeft)
            return -1;
        if (noTitleRight)
            return 1;
        return lhs.title.compareToIgnoreCase(rhs.title);
    };

    /**
     * Comparator that sorts {@link TaskItem}s by their due dates in ascending order. Default comparator.
     */
    private static Comparator<TaskItem> dueDateComparator = (lhs, rhs) -> {

        final boolean sameDay = lhs.dueDate != null && rhs.dueDate != null &&
                sdfDay.format(lhs.dueDate).equals(sdfDay.format(rhs.dueDate));
        final boolean sameMonth = lhs.dueDate != null && rhs.dueDate != null &&
                sdfDay.format(lhs.dueDate).substring(4, 6).equals(sdfDay.format(rhs.dueDate).substring(4, 6));

        int returnCode;

        if (lhs.dueDate != null) {
            // Compare non-null due dates, most recent ones at the top
            if (rhs.dueDate != null) {
                returnCode = lhs.dueDate.compareTo(rhs.dueDate);
            } else {
                // This task comes after the other task
                returnCode = 1;
            }
        } else if (rhs.dueDate != null) {
            // This task comes before the other task
            returnCode = -1;
        } else {
            // Both tasks have missing due dates, they are considered equal
            returnCode = 0;
        }

        if (sameDay) {
            if (returnCode == 0 || returnCode == -1)
                lhs.combineDay = true;
            else rhs.combineDay = true;
        }
        if (sameMonth) {
            lhs.combineMonth = true;
            rhs.combineMonth = false;
        } else {
            lhs.combineMonth = true;
            rhs.combineMonth = false;
        }

        return returnCode;
    };

    @Getter private final String taskId;
    @Getter private final String title;
    @Getter private final String notes;
    @Getter private final Date dueDate;
    @Getter private final Date completed;
    @Getter private final Date reminder;

    /**
     * Flag indicating that this task item should combine its
     * day (number & name) with the previous task item in the list
     */
    @Getter private boolean combineDay;

    /**
     * Flag indicating that this task item should combine its
     * month with the previous task item in the list
     */
    @Getter private boolean combineMonth;

    public TaskItem(TTask tTask) {
        title = tTask.getTitle();
        notes = tTask.getNotes();
        dueDate = tTask.getDue();
        completed = tTask.getCompleted();
        reminder = tTask.getReminder();
        taskId = tTask.getId();
    }

    @Override
    public int getType() {
        return R.id.task_item;
    }

    @Override
    public int getLayoutRes() {
        return R.layout.item_task;
    }

    @Override
    public ViewHolder getViewHolder(@NonNull View view) {
        return new ViewHolder(view);
    }

    /**
     * Compare by due date and return results in ascending order. <br>
     * Orders with missing due dates are considered high priority and they stay at the top
     */
    @Override
    public int compareTo(@NonNull TaskItem another) {
        return dueDateComparator.compare(this, another);
    }

    public static class ViewHolder extends FastAdapter.ViewHolder<TaskItem> {

        private static final int TOP_MARGIN_DP = 12;

        public final ItemTaskBinding binding;

        final int topMargin;

        final int left, right, top, bottom;

        ViewHolder(View view) {
            super(view);
            binding = DataBindingUtil.bind(view);

            // Set the reminder icon
            Drawable reminderIcon = VectorDrawableCompat.create(view.getResources(), R.drawable.ic_alarm_18dp, view.getContext().getTheme());
            binding.reminder.setCompoundDrawablesWithIntrinsicBounds(reminderIcon, null, null, null);

            // Calculate the top margin in px
            DisplayMetrics displayMetrics = view.getContext().getResources().getDisplayMetrics();
            topMargin = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, TOP_MARGIN_DP, displayMetrics);

            // Cache the padding
            left = binding.layoutTaskBody.getPaddingLeft();
            right = binding.layoutTaskBody.getPaddingRight();
            top = binding.layoutTaskBody.getPaddingTop();
            bottom = binding.layoutTaskBody.getPaddingBottom();
        }

        @Override
        public void bindView(@NonNull TaskItem item, @NonNull List<Object> payloads) {
            boolean combineDay = item.isCombineDay();
            boolean combineMonth = item.isCombineMonth();

            // Add the top padding for items that aren't combined
            binding.layoutTaskBody.setPadding(left,top + (!combineDay ? topMargin : 0), right, bottom);

            // Title
            binding.taskTitle.setText(item.getTitle());

            // Task description
            String notes = item.getNotes();
            if (notes == null) {
                binding.taskDescription.setVisibility(GONE);
            } else {
                binding.taskDescription.setText(notes);
                binding.taskDescription.setVisibility(VISIBLE);
            }

            // Reminder
            Date reminder = item.getReminder();
            if (reminder == null) {
                binding.reminder.setVisibility(GONE);
            } else {
                Timber.d("reminder is not null");
                binding.reminder.setText(DateUtils.formatTime(binding.getRoot().getContext(), reminder));
                binding.reminder.setVisibility(VISIBLE);
            }

            // Due date / Completed date
            Date completed = item.getCompleted();
            Date dueDate = item.getDueDate();

            if (combineDay) {
                binding.taskDate.setDate(null);
            } else {
                if (completed != null) {
                    binding.taskDate.setDate(completed);
                } else if (dueDate != null) {
                    binding.taskDate.setDate(dueDate);
                } else {
                    binding.taskDate.setDate(null);
                }
            }

            // Month
            if (combineMonth) {
                binding.month.setVisibility(GONE);
            } else {
                if (completed != null) {
                    binding.month.setText(DateUtils.getMonthAndYear(completed));
                } else if (dueDate != null) {
                    binding.month.setText(DateUtils.getMonthAndYear(dueDate));
                } else {
                    binding.month.setText("No due date");
                }
                binding.month.setVisibility(VISIBLE);
            }
        }

        @Override
        public void unbindView(@NonNull TaskItem item) {
            binding.layoutTaskBody.setPadding(0, 0, 0, 0);
            binding.taskTitle.setText(null);
            binding.taskDescription.setText(null);
            binding.reminder.setText(null);
            binding.taskDate.setDate(null);
        }
    }
}

File: t-tasks/src/main/java/com/teo/ttasks/ui/items/TaskItem.kt
package com.teo.ttasks.ui.items

import android.support.graphics.drawable.VectorDrawableCompat
import android.support.v7.widget.RecyclerView
import android.util.TypedValue
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import com.teo.ttasks.R
import com.teo.ttasks.data.model.Task
import com.teo.ttasks.databinding.ItemTaskBinding
import com.teo.ttasks.util.DateUtils
import com.teo.ttasks.util.DateUtils.Companion.sdfDay
import com.teo.ttasks.util.DateUtils.Companion.sdfMonth
import com.teo.ttasks.util.TaskType
import eu.davidea.flexibleadapter.FlexibleAdapter
import eu.davidea.flexibleadapter.items.AbstractFlexibleItem
import eu.davidea.flexibleadapter.items.IFlexible
import eu.davidea.viewholders.FlexibleViewHolder
import java.util.*


class TaskItem(task: Task,
               private val taskType: TaskType)
    : AbstractFlexibleItem<TaskItem.ViewHolder>(), Comparable<TaskItem> {

    val taskId: String = task.id

    val title: String = task.title!!

    val notes: String? = task.notes

    val dueDate: Date? = task.due

    val completed: Date? = task.completed

    val reminder: Date? = task.reminder

    private lateinit var viewHolder: ViewHolder

    val binding: ItemTaskBinding
        get() = viewHolder.binding

    /**
     * Flag indicating that this task item should combine its
     * day (number & name) with the previous task item in the list
     */
    private var combineDay: Boolean = false

    /**
     * Flag indicating that this task item should combine its
     * month with the previous task item in the list
     */
    private var combineMonth: Boolean = false

    override fun shouldNotifyChange(newItem: IFlexible<*>?): Boolean =
    // Should be bound again if ID is different
            taskId != (newItem as TaskItem).taskId

    override fun getItemViewType(): Int = taskType.id

    override fun getLayoutRes(): Int = R.layout.item_task

    override fun bindViewHolder(adapter: FlexibleAdapter<IFlexible<RecyclerView.ViewHolder>>, viewHolder: ViewHolder, position: Int, payloads: MutableList<Any?>) {
        this.viewHolder = viewHolder
        val binding = viewHolder.binding

        // Add the top padding for items that aren't combined
        binding.layoutTaskBody.setPadding(viewHolder.left,
                viewHolder.top + if (!combineDay) viewHolder.topMargin else 0,
                viewHolder.right,
                viewHolder.bottom)

        // Title
        binding.taskTitle.text = title

        // Task description
        if (notes == null) {
            binding.taskDescription.visibility = GONE
        } else {
            binding.taskDescription.text = notes
            binding.taskDescription.visibility = VISIBLE
        }

        // Reminder
        if (reminder == null) {
            binding.reminder.visibility = GONE
        } else {
            binding.reminder.text = DateUtils.formatTime(binding.root.context, reminder)
            binding.reminder.visibility = VISIBLE
        }

        // Due date / Completed date
        if (combineDay) {
            binding.taskDate.date = null
        } else {
            when {
                completed != null -> binding.taskDate.date = completed
                dueDate != null -> binding.taskDate.date = dueDate
                else -> binding.taskDate.date = null
            }
        }

        // Month
        if (combineMonth) {
            binding.month.visibility = GONE
        } else {
            when {
                completed != null -> binding.month.text = DateUtils.getMonthAndYear(completed)
                dueDate != null -> binding.month.text = DateUtils.getMonthAndYear(dueDate)
                else -> binding.month.setText(R.string.due_date_not_set)
            }
            binding.month.visibility = VISIBLE
        }
    }

    override fun createViewHolder(view: View, adapter: FlexibleAdapter<IFlexible<RecyclerView.ViewHolder>>): ViewHolder {
        return ViewHolder(view, adapter)
    }

    /**
     * Compare by due date and return results in ascending order. <br></br>
     * Orders with missing due dates are considered high priority and they stay at the top
     */
    override fun compareTo(other: TaskItem): Int = dueDateComparator.compare(this, other)

    override fun equals(other: Any?): Boolean =
            other is TaskItem &&
                    taskId == other.taskId &&
                    title == other.title &&
                    notes == other.notes &&
                    dueDate == other.dueDate &&
                    completed == other.completed &&
                    reminder == other.reminder

    override fun hashCode(): Int {
        var result = taskId.hashCode()
        result = 37 * result + title.hashCode()
        notes?.let { result = 37 * result + it.hashCode() }
        dueDate?.let { result = 37 * result + it.hashCode() }
        completed?.let { result = 37 * result + it.hashCode() }
        reminder?.let { result = 37 * result + it.hashCode() }
        return result
    }

    class ViewHolder internal constructor(view: View, adapter: FlexibleAdapter<*>) : FlexibleViewHolder(view, adapter) {

        val binding = ItemTaskBinding.bind(view)

        internal val topMargin: Int

        internal val left: Int
        internal val right: Int
        internal val top: Int
        internal val bottom: Int

        init {

            // Set the reminder icon
            val reminderIcon = VectorDrawableCompat.create(view.resources, R.drawable.ic_alarm_18dp, view.context.theme)
            binding.reminder.setCompoundDrawablesWithIntrinsicBounds(reminderIcon, null, null, null)

            // Calculate the top margin in px
            val displayMetrics = view.context.resources.displayMetrics
            topMargin = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, TOP_MARGIN_DP.toFloat(), displayMetrics).toInt()

            // Cache the padding
            left = binding.layoutTaskBody.paddingLeft
            right = binding.layoutTaskBody.paddingRight
            top = binding.layoutTaskBody.paddingTop
            bottom = binding.layoutTaskBody.paddingBottom
        }

        companion object {
            private const val TOP_MARGIN_DP = 12
        }
    }

    companion object {

        /**
         * Comparator that sorts [TaskItem]s by their completion date in descending order.
         */
        var completionDateComparator = Comparator<TaskItem> { lhs, rhs ->

            val sameDay = sdfDay.format(lhs.completed) == sdfDay.format(rhs.completed)
            val sameMonth = sdfMonth.format(lhs.completed) == sdfMonth.format(rhs.completed)

            // Check the completed date
            val returnCode =
                    if (lhs.completed == null && rhs.completed == null) {
                        0
                    } else if (lhs.completed == null) {
                        1
                    } else if (rhs.completed == null) {
                        -1
                    } else {
                        rhs.completed.compareTo(lhs.completed)
                    }

            // Decide whether the day number and name will be shown i.e. whether the task item will be combined or not
            if (sameDay) {
                if (returnCode > 0) {
                    // rhs comes after lhs
                    lhs.combineDay = true
                } else {
                    // lhs comes after rhs
                    rhs.combineDay = true
                }
            }

            // Decide whether the month should be shown
            if (sameMonth) {
                if (returnCode > 0) {
                    // rhs comes after lhs
                    lhs.combineMonth = true
                } else {
                    // lhs comes after rhs
                    rhs.combineMonth = true
                }
            }

            return@Comparator returnCode
        }

        var alphabeticalComparator = Comparator<TaskItem> { lhs, rhs ->
            // Titles should never be null or empty but just in case
            val noTitleLeft = lhs.title.isEmpty()
            val noTitleRight = rhs.title.isEmpty()
            if (noTitleLeft && noTitleRight)
                return@Comparator 0
            if (noTitleLeft)
                return@Comparator -1
            if (noTitleRight)
                return@Comparator 1
            return@Comparator lhs.title.compareTo(rhs.title, ignoreCase = true)
        }

        /**
         * Comparator that sorts [TaskItem]s by their due dates in ascending order. Default comparator.
         */
        private val dueDateComparator = Comparator<TaskItem> { lhs, rhs ->

            val sameDay = lhs.dueDate != null && rhs.dueDate != null &&
                    sdfDay.format(lhs.dueDate) == sdfDay.format(rhs.dueDate)
            val sameMonth = lhs.dueDate != null && rhs.dueDate != null &&
                    sdfDay.format(lhs.dueDate).substring(4, 6) == sdfDay.format(rhs.dueDate).substring(4, 6)

            val returnCode =
                    if (lhs.dueDate != null) {
                        // Compare non-null due dates, most recent ones at the top
                        if (rhs.dueDate != null) {
                            lhs.dueDate.compareTo(rhs.dueDate)
                        } else {
                            // This task comes after the other task
                            1
                        }
                    } else if (rhs.dueDate != null) {
                        // This task comes before the other task
                        -1
                    } else {
                        // Both tasks have missing due dates, they are considered equal
                        0
                    }

            if (sameDay) {
                if (returnCode == 0 || returnCode == -1)
                    lhs.combineDay = true
                else
                    rhs.combineDay = true
            }
            if (sameMonth) {
                lhs.combineMonth = true
                rhs.combineMonth = false
            } else {
                lhs.combineMonth = true
                rhs.combineMonth = false
            }

            return@Comparator returnCode
        }
    }
}
